File: ProblemSolve/1048.cs
/************************
 * Longest String Chain *
 ************************/

public class Solution {
    public int LongestStrChain(string[] words) {
        int ans = 1;
        Dictionary<string, int> dp = new Dictionary<string, int>();
        Array.Sort(words, (a, b) => {return a.Length < b.Length ? -1 : 1;});
        
        foreach(string word in words){
            if(!dp.ContainsKey(word)){
                dp.Add(word, 1);
            }
            
            for(int i=0; i<word.Length; ++i){
                string key = word.Substring(0, i) + word.Substring(i+1);
            
                if(dp.ContainsKey(key)){
                    dp[word] = dp[key] + 1;
                    ans = Math.Max(ans, dp[word]);
                }
            }
        }
        
        return ans;
    }
}

File: ProblemSolve/1048.cpp
/************************
 * Longest String Chain *
 ************************/

class Solution {
public:
    int longestStrChain(vector<string>& words) {
        int ans = 1;
        sort(words.begin(), words.end(), [](string& a, string& b){return a.size() < b.size();});
        unordered_map<string, int> dp;
        
        for(auto& word : words){
            if(dp.find(word) == dp.end()){
                dp[word] = 1;
            }
            
            for(int i=0; i<word.size(); ++i){
                string key = word.substr(0, i) + word.substr(i+1);
                
                if(dp.find(key) != dp.end()){
                    dp[word] = dp[key] + 1;
                    ans = max(ans, dp[word]);
                }
            }
        }
        
        return ans;
    }
};